### ISSUE TITLE
codegen: Order of field arguments (API consistency)

### ISSUE THREAD

**Issue by helderco** [dagger/dagger#3554]
Feedback from user _talios_ on Discord: [Principle of Least Surprise - API Consistency](https://discord.com/channels/707636530424053791/1034775679897718834)

---

On codegen we're sorting everything (types, fields and arguments) to produce a deterministic output. Specifically with field arguments for example, it seems that `graphql-go` is handling them in a map which loses their order.

The problem is we can get inconsistencies in the generated client:

```graphql
# container.graphqlps
type Container {
    withMountedFile(path: String!, source: FileID!): Container!
    withMountedCache(path: String!, cache: CacheID!, source: DirectoryID): Container!
}
```
```go
// api.gen.go
func (r *Container) WithMountedFile(path string, source FileID) *Container
func (r *Container) WithMountedCache(cache CacheID, path string, opts ...ContainerWithMountedCacheOpts) *Container
```

As you can see, the `cache` in `withMountedCache` gets sorted to the start.

In Python we can require all arguments to be keyword based but not all SDKs can do this and I think it's good to allow positional:

```python
api.container().from_("alpine")
# vs
api.container().from_(address="alpine")
```

Question is, do we have any idea on how to better control the order of the arguments?

- Patch `graphql-go` or propose change upstream?
- Annotate schema?
- Weighted sorting based on type (e.g., custom scalars vs primitives)?
- Have a resolver to query these argument orders and add it to the introspection query or to a separate query to help the codegen sort based on that?

Ideally we would be able to use the schema order as the source of truth and replicated it because no matter what automatic sorting logic you use, it may always make some other sense that's more obvious for usage when you're crafting the schema.

**Comment by aluzzardi** [comment]
Merge from #3589

Field ordering in codegen is alphabetical.

This causes inconsistencies, e.g.:

https://discord.com/channels/707636530424053791/1034775679897718834

```go
maven = maven.WithMountedCache(cacheID, "/root/.m2/repository")
maven = maven.WithMountedFile("/root/.m2/settings.xml", mavenSettingsId)
```

**Proposal: Keep same field ordering as defined in the schema**

This is actually tricky to do because the introspection query returns a random order.

Solutions:

1) Patch `graphql-go` so schema field ordering is preserved in the introspection API (**preferred solution**)
2) Change codegen to no longer use introspection query. Instead, parse the schema directly, hopefully keeping the same order. Requires adding an API endpoint to retrieve the raw schema (e.g. `query { _sdl }`)

/cc @vito @sipsma @helderco

**Comment by aluzzardi** [comment]
I patched upstream and opened a PR: https://github.com/graphql-go/graphql/pull/654

However, it's a breaking API change so the changes of it being merged are slim.

So we either:
- Use a fork of graphql-go with this change
- Find another solution

**Comment by talios** [comment]
@aluzzardi would using something like https://chromium.googlesource.com/external/github.com/emirpasic/gods/+/refs/heads/master/README.md#linkedhashmap be a cleaner change (altho that is an additional dependency).

The `linkedHashMap` preserves insertion order, but appears to implement the standard map interfaces, so may be a less invasive PR.

**Comment by aluzzardi** [comment]
So the problem is in `graphql-go`, that "map" is part of the user API to define the schema (e.g. it's not an internal thing nobody sees), like so:

```go
var queryType = graphql.NewObject(
	graphql.ObjectConfig{
		Name: "Query",
		Fields: graphql.Fields{
			"user": &graphql.Field{
				Type: userType,
				Args: graphql.FieldConfigArgument{
					"id": &graphql.ArgumentConfig{
						Type: graphql.String,
					},
				},
```

The problem is `"id": &graphql.ArgumentConfig{` --> random order.

If we were to use a library to define args, then that library would become part of the public API of graphql-go :/

**Comment by aluzzardi** [comment]
Fixed by #3631
